### PR TITLE
[InstCombine] Handle IsInf/IsZero idioms

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch does the following folds:
```
icmp eq/ne (bitcast X to int), (bitcast +/-inf to int) -> llvm.is.fpclass(X, (~)fcPosInf/fcNegInf)
icmp eq/ne (bitcast X to int), (bitcast +0/-0 to int) -> llvm.is.fpclass(X, (~)fcPosZero/fcNegZero)
```
Alive2: https://alive2.llvm.org/ce/z/JJmEE9

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll [hunk 1]
```diff
@@ -200,4 +312,44 @@ define i1 @f32_fcnan_fcinf_noimplicitfloat(float %a) #0 {
...
+; CHECK-NEXT:    call void @usei32(i32 [[I32]])
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq i32 [[I32]], 2139095040
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %i32 = bitcast float %a to i32
+  call void @usei32(i32 %i32)
+  %cmp = icmp eq i32 %i32, 2139095040
+  ret i1 %cmp
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:344]: This doesn't include the fpext/fptrunc cases

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -3261,6 +3261,20 @@ Instruction *InstCombinerImpl::foldICmpBitCast(ICmpInst &Cmp) {
                                 ConstantInt::getAllOnesValue(NewType));
         }
       }
+
+      // icmp eq/ne (bitcast X to int), special fp -> llvm.is.fpclass(X, class)
+      Type *FPType = SrcType->getScalarType();
+      if (!Cmp.getParent()->getParent()->hasFnAttribute(
+              Attribute::NoImplicitFloat) &&
+          Cmp.isEquality() && FPType->isIEEELikeFPTy()) {
+        unsigned Mask = APFloat(FPType->getFltSemantics(), *C).classify();
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3270]: I believe classify returns the true FPClassTest, you don't need the implicit cast to unsigned

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -3261,6 +3261,20 @@ Instruction *InstCombinerImpl::foldICmpBitCast(ICmpInst &Cmp) {
...
+        if (Mask & (fcInf | fcZero)) {
+          if (Pred == ICmpInst::ICMP_NE)
+            Mask = ~Mask & fcAllFlags;
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:3273]: You don't need the clamp to fcAllFlags, it implicitly happens with the bit mask enum type

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:344]: Sorry, I cannot get your point. As all the tests are bit operations on floating point numbers, I don't think this patch is related to fpext/fptrunc.

Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:344]: I mean I see match(BCSrcOp, m_FPExt(m_Value(X))) here

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:344]: Oh, it exists before this patch.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp (hunk 2 above)
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Done.

**Comment by dtcxzyw** [comment]
Ping.

**Review by arsenm — APPROVED** [review]

**Pull request merged**